lib/util/cleverMerge.js

```diff
@@ -74,6 +74,122 @@ const cachedSetProperty = (obj, property, value) => {
 	return result;
 };
 
+/**
+ * @typedef {Object} ObjectParsedPropertyEntry
+ * @property {any | undefined} base base value
+ * @property {string | undefined} byProperty the name of the selector property
+ * @property {Map<string, any>} byValues value depending on selector property, merged with base
+ */
+
+/**
+ * @param {any} obj the object
+ * @returns {Map<string, ObjectParsedPropertyEntry} parsed properties
+ */
+const parseObject = obj => {
+	const info = new Map();
+	const getInfo = p => {
+		const entry = info.get(p);
+		if (entry !== undefined) return entry;
+		const newEntry = {
+			base: undefined,
+			byProperty: undefined,
+			byValues: undefined
+		};
+		info.set(p, newEntry);
+		return newEntry;
+	};
+	for (const key of Object.keys(obj)) {
+		if (key.startsWith("by")) {
+			const byProperty = key;
+			const byObj = obj[byProperty];
+			for (const byValue of Object.keys(byObj)) {
+				const obj = byObj[byValue];
+				for (const key of Object.keys(obj)) {
+					const entry = getInfo(key);
+					if (entry.byProperty === undefined) {
+						entry.byProperty = byProperty;
+						entry.byValues = new Map();
+					} else if (entry.byProperty !== byProperty) {
+						throw new Error(
+							`${byProperty} and ${entry.byProperty} for a single property is not supported`
+						);
+					}
+					entry.byValues.set(byValue, obj[key]);
+					if (byValue === "default") {
+						for (const otherByValue of Object.keys(byObj)) {
+							if (!entry.byValues.has(otherByValue))
+								entry.byValues.set(otherByValue, undefined);
+						}
+					}
+				}
+			}
+		} else {
+			const entry = getInfo(key);
+			entry.base = obj[key];
+		}
+	}
+	return info;
+};
+
+/**
+ * @param {Map<string, ObjectParsedPropertyEntry>} info property entries
+ * @returns {any} the object
+ */
+const serializeObject = info => {
+	const obj = {};
+	// Setup byProperty structure
+	for (const entry of info.values()) {
+		if (entry.byProperty !== undefined) {
+			const byObj = (obj[entry.byProperty] = obj[entry.byProperty] || {});
+			for (const byValue of entry.byValues.keys()) {
+				byObj[byValue] = byObj[byValue] || {};
+			}
+		}
+	}
+	for (const [key, entry] of info) {
+		if (entry.base !== undefined) {
+			obj[key] = entry.base;
+		}
+		// Fill byProperty structure
+		if (entry.byProperty !== undefined) {
+			const byObj = (obj[entry.byProperty] = obj[entry.byProperty] || {});
+			for (const byValue of Object.keys(byObj)) {
+				const value = getFromByValues(entry.byValues, byValue);
+				if (value !== undefined) byObj[byValue][key] = value;
+			}
+		}
```

Comment:
Small optimization
```suggestion
	for (const [key, entry] of info) {
		if (entry.base !== undefined) {
			obj[key] = entry.base;
		}
		// Fill byProperty structure
		if (entry.byProperty !== undefined) {
			const byObj = (obj[entry.byProperty] = obj[entry.byProperty] || {});
			const keys = entry.byValues.keys();
			for (const byValue of keys)) {
			    const byValueObj = (byObj[byValue] = byObj[byValue] || {});
				const value = getFromByValues(entry.byValues, byValue);
				if (value !== undefined) byValueObj[key] = value;
			}
		}
```

Reply:
I actually had this before, but there is an intentional different here: 

``` diff
+for (const byValue of Object.keys(byObj)) {
-for (const byValue of entry.byValues.keys())) {
```

We need to insert the default value for all generated byValues (entry.byValues may not cover all of them). So to do that we need to walk over all `info` values to get all possible byValues. It would also possible to store them in a `Set` and walk them later, but for performance reasons I directly create the structure and kind of store them in `Object.keys(byObj)`. This also allow to skip conditional structure creation (`obj[entry.byProperty] = obj[entry.byProperty] || {}`) later.